Recursion limit tripped in doc but not in other profiles

The diesel benchmark has been broken in some recent commit: https://perf.rust-lang.org/status.html.

ref: [zulip comment](https://rust-lang.zulipchat.com/#narrow/stream/247081-t-compiler.2Fperformance/topic/.60diesel.60.20benchmark.20broken/near/295189845)

The error seems to be caused https://github.com/rust-lang/rust/issues/100757 (cc @ouz-a). Curiously, this only happens in `doc`. 

Here's the error:

```
 Documenting diesel v1.4.8 (/home/rylevick/code/rustc-perf/collector/benchmarks/diesel-1.4.8)
error[E0275]: overflow evaluating the requirement `<_ as query_source::Column>::Table`
    |
    = help: consider increasing the recursion limit by adding a `#![recursion_limit = "256"]` attribute to your crate (`diesel`)
note: required for `&expression::operators::Eq<_, _>` to implement `insertable::Insertable<_>`
   --> src/expression/operators.rs:376:21
    |
376 | impl<'a, T, Tab, U> Insertable<Tab> for &'a Eq<T, U>
```

The line where the error triggers is:

```rust
impl<'a, T, Tab, U> Insertable<Tab> for &'a Eq<T, U> // <--- HERE
where
    T: Copy,
    Eq<T, &'a U>: Insertable<Tab>,
{
    type Values = <Eq<T, &'a U> as Insertable<Tab>>::Values;

    fn values(self) -> Self::Values {
        Eq::new(self.left, &self.right).values()
    }
}
```

You can reproduce this issue by:
* cloning the rustc-perf repo
* going to diesel-1.4.8 benchmark in that repo
* installing commit `a72b0e79ba45f05b339b45034996086bf10d3c26 ` using `rustup-toolchain-install-master` 
* running `cargo +a72b0e79ba45f05b339b45034996086bf10d3c26 doc`

--- a/compiler/rustc_trait_selection/src/traits/project.rs
+++ b/compiler/rustc_trait_selection/src/traits/project.rs
@@ -557,7 +557,10 @@ impl<'a, 'b, 'tcx> TypeFolder<'tcx> for AssocTypeNormalizer<'a, 'b, 'tcx> {
                 // For cases like #95134 we would like to catch overflows early
                 // otherwise they slip away away and cause ICE.
                 let recursion_limit = self.tcx().recursion_limit();
-                if !recursion_limit.value_within_limit(self.depth) {
+                if !recursion_limit.value_within_limit(self.depth)
+                    // HACK: Don't overflow when running cargo doc see #100991
+                    && !self.tcx().sess.opts.actually_rustdoc
+                {
                     let obligation = Obligation::with_depth(
                         self.cause.clone(),
                         recursion_limit.0,

